src/archive/src/main.rs

```diff
@@ -0,0 +1,394 @@
+use candid::{CandidType, Deserialize, Principal};
+use ic_cdk::api::stable::stable64_size;
+use ic_cdk::api::time;
+use ic_cdk::{caller, trap};
+use ic_cdk_macros::{init, post_upgrade, query, update};
+use ic_stable_structures::memory_manager::{MemoryId, MemoryManager, VirtualMemory};
+use ic_stable_structures::{
+    cell::Cell as StableCell, log::Log, DefaultMemoryImpl, Memory as StableMemory,
+    RestrictedMemory, StableBTreeMap, Storable,
+};
+use internet_identity_interface::*;
+use metrics_encoder::MetricsEncoder;
+use serde_bytes::ByteBuf;
+use std::borrow::Cow;
+use std::cell::RefCell;
+
+type Memory = RestrictedMemory<DefaultMemoryImpl>;
+type StableLog = Log<VirtualMemory<Memory>, VirtualMemory<Memory>>;
+type ConfigCell = StableCell<ArchiveConfig, Memory>;
+type UserIndex = StableBTreeMap<VirtualMemory<Memory>, UserIndexKey, ()>;
+
+const GIB: u64 = 1 << 30;
+const WASM_PAGE_SIZE: u64 = 65536;
+/// The maximum number of Wasm pages that we allow to use for the stable storage.
+const MAX_WASM_PAGES: u64 = 8 * GIB / WASM_PAGE_SIZE;
+
+const LOG_INDEX_MEMORY_ID: MemoryId = MemoryId::new(0);
+const LOG_DATA_MEMORY_ID: MemoryId = MemoryId::new(1);
+const USER_INDEX_MEMORY_ID: MemoryId = MemoryId::new(2);
+
+/// Length of the user index key. Changing this value requires a stable memory migration.
+const USER_INDEX_KEY_LENGTH: usize = 24;
```

Comment:
out of curiosity, where does `24` come from?

Reply:
It is the size of the byte representation of the `AnchorIndexKey` struct. But I have replaced it now with `std::mem::size_of::<AnchorIndexKey>()`.